packages/migrate/migrations/svelte-4/migrate.js

```diff
@@ -1,39 +1,113 @@
-		(_match, attrs, contents, whitespace) => {
-			return `<script${attrs}>${transform_code(contents)}</script>${whitespace}`;
+export function update_pkg_json() {
+	fs.writeFileSync(
+		'package.json',
+		update_pkg_json_content(fs.readFileSync('package.json', 'utf8'))
+	);
+}
+
+/**
+ * @param {string} content
+ */
+export function update_pkg_json_content(content) {
+	const indent = content.split('\n')[1].match(/^\s+/)?.[0] || '  ';
+	const pkg = JSON.parse(content);
+
+	/**
+	 * @param {string} name
+	 * @param {string} version
+	 * @param {string} [additional]
+	 */
+	function update_pkg(name, version, additional = '') {
+		if (pkg.dependencies[name]) {
+			log_migration(`Updated ${name} to ${version}${additional}`);
+			pkg.dependencies[name] = version;
 		}
+		if (pkg.devDependencies[name]) {
+			log_migration(`Updated ${name} to ${version}${additional}`);
+			pkg.devDependencies[name] = version;
+		}
+	}
+
+	update_pkg('svelte', '^4.0.0');
+	update_pkg('svelte-check', '^3.4.3');
+	update_pkg('svelte-preprocess', '^5.0.3');
+	update_pkg('@sveltejs/kit', '^1.20.4');
+	update_pkg('@sveltejs/vite-plugin-svelte', '^2.4.1');
+	update_pkg(
+		'svelte-loader',
+		'^3.1.8',
+		' (if you are still on webpack 4, you need to update to webpack 5)'
+	);
+	update_pkg('rollup-plugin-svelte', '^7.1.5');
+	update_pkg('prettier-plugin-svelte', '^2.10.1');
+	update_pkg('eslint-plugin-svelte', '^2.30.0');
+	update_pkg(
+		'typescript',
+		'^5.0.0',
+		' (this might introduce new type errors due to breaking changes within TypeScript)'
 	);
-	fs.writeFileSync(file_path, transform_svelte_code(updated), 'utf-8');
+
+	return JSON.stringify(pkg, null, indent);
+}
+
+/**
+ * @param {string} file_path
+ * @param {boolean} migrate_transition
+ */
+export function update_svelte_file(file_path, migrate_transition) {
+	try {
+		const content = fs.readFileSync(file_path, 'utf-8');
+		const updated = content.replace(
+			/<script([^]*?)>([^]+?)<\/script>(\n*)/g,
+			(_match, attrs, contents, whitespace) => {
+				return `<script${attrs}>${transform_code(
+					contents,
+					(attrs.includes('lang=') || attrs.includes('type=')) &&
+						(attrs.includes('ts') || attrs.includes('typescript'))
+				)}</script>${whitespace}`;
+			}
+		);
+		fs.writeFileSync(file_path, transform_svelte_code(updated, migrate_transition), 'utf-8');
+	} catch (e) {
+		console.error(`Error updating ${file_path}: ${/** @type {any} */ (e)?.message}`);
+	}
 }
 
 /** @param {string} file_path */
 export function update_js_file(file_path) {
-	const content = fs.readFileSync(file_path, 'utf-8');
-	const updated = transform_code(content);
-	fs.writeFileSync(file_path, updated, 'utf-8');
+	try {
+		const content = fs.readFileSync(file_path, 'utf-8');
+		const updated = transform_code(content, file_path.endsWith('.ts'));
+		fs.writeFileSync(file_path, updated, 'utf-8');
+	} catch (e) {
+		console.error(`Error updating ${file_path}: ${/** @type {any} */ (e)?.message}`);
```

Comment:
```suggestion
		console.error(`Error updating ${file_path}:`, e);
```

I think this will get better formatting and probably nice to have the stack trace too.